File: sln/bmaclient/LTLPolarityCheckRole/WorkerRole.fs
namespace LTLPolarityCheckRole

open System
open System.Collections.Generic
open System.Diagnostics
open System.Linq
open System.Net
open System.Threading
open Microsoft.WindowsAzure
open Microsoft.WindowsAzure.Diagnostics
open Microsoft.WindowsAzure.ServiceRuntime
open Microsoft.WindowsAzure.Storage
open Newtonsoft.Json

open bma.Cloud
open bma.LTLPolarity

type WorkerRole() =
    inherit RoleEntryPoint() 

    let log message (kind : string) = Trace.TraceInformation(message, kind)
    let logInfo message = log message "Information"

    let mutable worker : IWorker option = None

    let doJob(jobId: Guid, input: IO.Stream) : IO.Stream = 
        logInfo (sprintf "Doing the job #%O" jobId)

        let reader = new IO.StreamReader(input)
        let input_s = reader.ReadToEnd()
        let query = JsonConvert.DeserializeObject<LTLPolarityAnalysisInputDTO>(input_s)

        
        let res = bma.LTLPolarity.Algorithms.Check(query)



        let jsRes = JsonConvert.SerializeObject(res)
        let output_s = jsRes.ToString()
        let ms = new IO.MemoryStream()
        let writer = new IO.StreamWriter(ms)
        writer.Write(output_s)
        writer.Flush()
        ms.Position <- 0L
        upcast ms


    override wr.Run() =
        let worker = worker.Value

        logInfo "LTLPolarityCheckRole entry point called"
        worker.Start(doJob, TimeSpan.FromSeconds 1.0)
        base.Run()

    override wr.OnStart() = 
        let storageAccount = CloudStorageAccount.Parse(RoleEnvironment.GetConfigurationSettingValue("StorageConnectionString"));
        let schedulerName = "ltlpolarity"; // todo: can differ for different controllers; use setter injection with name?

        ServicePointManager.DefaultConnectionLimit <- 12

        try
            worker <- Worker.Create (storageAccount, schedulerName) |> Some
            base.OnStart()
        with
        | ex -> 
            log (sprintf "Exception during OnStart: %O" ex) "Error"
            false

    override wr.OnStop() =
        worker |> Option.iter (fun w -> w.Dispose())

File: src/LTLCheckRole/WorkerRole.cs
// Copyright (c) Microsoft Research 2016
// License: MIT. See LICENSE
using bma.Cloud;
using Microsoft.WindowsAzure.ServiceRuntime;
using Microsoft.WindowsAzure.Storage;
using Newtonsoft.Json;
using System;
using System.Diagnostics;
using System.IO;
using System.Net;
using System.Threading;

namespace LTLCheckRole
{
    public class WorkerRole : RoleEntryPoint
    {
        private readonly CancellationTokenSource cancellationTokenSource = new CancellationTokenSource();
        private readonly ManualResetEvent runCompleteEvent = new ManualResetEvent(false);
        
        private IWorker worker;
        private string basePath;

        public override void Run()
        {
            Trace.TraceInformation("LTLCheckRole is running");

            try
            {
                basePath = Environment.CurrentDirectory;
                worker.Process(DoJob, TimeSpan.FromSeconds(0.5), TimeSpan.FromMinutes(1.0));
            }
            finally
            {
                this.runCompleteEvent.Set();
            }
        }

        public override bool OnStart()
        {
            // Set the maximum number of concurrent connections
            ServicePointManager.DefaultConnectionLimit = 12;

            // For information on handling configuration changes
            // see the MSDN topic at http://go.microsoft.com/fwlink/?LinkId=166357.

            bool result = base.OnStart();

            var storageAccount = CloudStorageAccount.Parse(RoleEnvironment.GetConfigurationSettingValue("StorageConnectionString"));
            var schedulerName = "ltlpolarity"; // todo: can differ for different controllers; use setter injection with name?

            try
            {
                if (result)
                {
                    var settings = new WorkerSettings(TimeSpan.FromHours(1), 3, TimeSpan.FromMinutes(5), TimeSpan.FromMinutes(1));
                    worker = Worker.Create(storageAccount, schedulerName, settings);
                    Trace.TraceInformation("LTLCheckRole has been started");
                }
            }
            catch (Exception ex)
            {
                Trace.TraceError("Exception during OnStart: {0}", ex);
                result = false;
            }

            return result;
        }

        public override void OnStop()
        {
            Trace.TraceInformation("LTLCheckRole is stopping");

            this.cancellationTokenSource.Cancel();
            this.runCompleteEvent.WaitOne();

            base.OnStop();

            if (worker != null)
                worker.Dispose();

            Trace.TraceInformation("LTLCheckRole has stopped");
        }
        

        private Stream DoJob(Guid jobId, Stream input)
        {
            Trace.TraceInformation("Doing the job {0}", jobId);
            var sw = new Stopwatch();
            sw.Start();

            string input_s = (new StreamReader(input)).ReadToEnd();

            // JobRunner kills the process, if this thread is aborted due to job cancellation.
            JobsRunner.JobResult result = JobsRunner.Job.RunToCompletion(Path.Combine(basePath, "AnalyzeLTL.exe"), input_s, -1);

            var ms = new MemoryStream();
            var writer = new StreamWriter(ms);
            writer.Write(result.Content);
            writer.Flush();
            ms.Position = 0;

            sw.Stop();
            Trace.TraceInformation("Job {0} done ({1})", jobId, sw.Elapsed);

            return ms;
        }
    }
}
